File: project6/submit/Go/task2.go
/*
  CS333 Project 6
  Part II: Selected Language - Go
  Task 2
  Show how Go reads from and writes to a file

  Yixuan Qiu
  Apr 18 2020
*/

package main

import (
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"net/http"
	"os"
)

// check errors
func check(e error) {
	if e != nil {
		panic(e)
	}
}

func main() {
	// 1) open, close, and read from a file
	// open a file
	f, err := os.Open("dat1") // support for binary file
	if err != nil {
		fmt.Println(err)
		return
	}

	// close a file
	defer f.Close()

	// allow up to 4096 bytes to be read
	data := make([]byte, 4096)
	for {
		data = data[:cap(data)]
		// 2) io.Reader interface's Read method
		n, err := f.Read(data)
		if err != nil {
			if err == io.EOF {
				break
			}
			fmt.Println(err)
			return
		}
		data = data[:n]
	}

	// write a string into a file
	d1 := []byte("hello\ngo\n")
	err := ioutil.WriteFile("dat1", d1, 0644)
	check(err)
	// check the content of the written file in terminal:
	// cat dat1

	// 3) open file URL
	response, err := http.Get("https://raw.githubusercontent.com/qiuyixuan/CS333/master/test.txt")

	if err != nil {
		fmt.Println(err)
		return
	}

	defer response.Body.Close()

	// Copy data from the response to standard output
	n, err1 := io.Copy(os.Stdout, response.Body) //use package "io" and "os"
	if err != nil {
		fmt.Println(err1)
		return
	}

	fmt.Println("\nNumber of bytes:", n)

	// open a file on the local disk
	file, err := os.Open("test.txt")
	// if an error occurs
	if err != nil {
		// call log.Fatal to print the error message and stop
		log.Fatal(err)
	}
	defer file.Close()
	b, err := ioutil.ReadAll(file)
	fmt.Print("Content of file:\n", string(b), "\n")

	// 4) user input information into the program interactively
	// Read integer
	var i int
	fmt.Println("Please enter an integer:")
	fmt.Scan(&i)
	fmt.Println("Integer entered: ", i)

	//  Read string
	var str string
	fmt.Println("\nPlease enter a word:")
	fmt.Scan(&str)
	fmt.Println("Word entered: ", str)

}

File: project6/submit/JavaScript/task2.js
/**
 * CS333 Project 6
 * task1.js
 * File I/O in JS: open, close, read file and binary finle; 
 * Reading from URLs;
 * Interactive: input from user
 * 
 * Di Luo
 * 05/10/2020
 */

function readtext(){
    const fs = require('fs');
    fs.open('wctest.txt','r+',function(err,fd){
        if (err){
            return console.error(err);
        }
        var t = fs.readFileSync('wctest.txt',"utf-8");
        console.log(t);
        fs.closeSync(fd);
    });
}

function write(){
    const fs = require('fs');
    fs.writeFile('file.txt',"Hello World!", function(err){
        if(err){
            return console.error(err);
        }
    });
}

function readbinary(){
    var fs = require('fs');
    var b = fs.readFileSync("wctest.txt", "binary");
}

function interactive(){
    const readline = require('readline').createInterface({
        input: process.stdin,
        output: process.stdout
    });
    readline.question("What's your name?", (name)=> {
        console.log(`Hi ${name}! Welcome!`)
        readline.close
    })
}
// readtext()
// write()
// readbinary()
interactive();